src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc

```diff
@@ -441,6 +455,66 @@ TEST_F(GcsTaskManagerTest, TestGetTaskEventsByJob) {
                      reply_job2.mutable_events_by_task());
 }
 
+TEST_F(GcsTaskManagerTest, TestIntermediateTerminatedChild) {
+  // Test the below case:
+  // A -> B -> C -> D -> E
+  //  E: running at 1
+  //  D: fails at 2
+  //  C: running at 3
+  //  B: finishes at 4
+  //  A: failed at 5
+  //
+  // Expect D and E fails at 2, A and C fails at 5, B finishes at 4.
+  auto task_ids = GenTaskIDs(5);
+  auto A = task_ids[0];
+  auto B = task_ids[1];
+  auto C = task_ids[2];
+  auto D = task_ids[3];
+  auto E = task_ids[4];
+
+  // E running at 1
+  SyncAddTaskEvent({E}, {{rpc::TaskStatus::RUNNING, 1}}, D);
+
+  // D fails at 2
+  SyncAddTaskEvent({D}, {{rpc::TaskStatus::FAILED, 2}}, C);
+
+  // C running at 3
+  SyncAddTaskEvent({C}, {{rpc::TaskStatus::RUNNING, 3}}, B);
+
+  // B finishes at 4
+  SyncAddTaskEvent({B}, {{rpc::TaskStatus::FINISHED, 4}}, A);
+
+  // A fails at 5
+  SyncAddTaskEvent({A}, {{rpc::TaskStatus::FAILED, 5}});
```

Comment:
can you also add a test case where 

1. parent task failure reported
2. child task reported (as non failure state)
3. Make sure both parent & child still fails

Reply:
Yeah, I believe this is covered in the `TestFailedParentShouldFailGrandChildren` below?